### ISSUE TITLE
Segmentation fault: Maximum number of clients reached

### ISSUE THREAD

**Issue by hakostra** [pyvista/pyvista#1180]
**Describe the bug, what's wrong, and what you expect:**

When creating a sequence of visualizations, the memory usage by the Python process keep rising, indicating that there are some resources not being deallocated properly, before it eventually segfault. The number of plots you can make before it segfaults seems to be around the same number each time, just above 180... The complexity of the images does not seem to affect, i.e. I can create the same number of figures before the segfault, independent on the complexity of the figure.

The segfault message is:
```
Maximum number of clients reachedFatal Python error: Segmentation fault

Current thread 0x00007f7b2539d740 (most recent call first):
  File "/home/hakostra/tmp/vtktest/lib/python3.8/site-packages/pyvista/plotting/plotting.py", line 841 in render
  File "/home/hakostra/tmp/vtktest/lib/python3.8/site-packages/pyvista/plotting/plotting.py", line 4266 in show
  File "pyvista-leak.py", line 17 in slice_once
  File "pyvista-leak.py", line 42 in <module>
Segmentation fault (core dumped)
```

-----

**To Reproduce**

Please include a code snippet to reproduce the bug in the block below:

```py
def slice_once(grid, x0, nvec):
    myslice = grid.slice(origin=x0, normal=nvec)

    plotter = pv.Plotter(off_screen=True, window_size=[1920, 1200],
                         lighting='none')
    plotter.add_mesh(myslice)
    plotter.enable_parallel_projection()
    plotter.camera_position = "yz"
    plotter.show(screenshot='demo.png')


print(pv.Report())

rng = np.random.default_rng()
size = (14, 14, 14,)

# Grid 1
x = np.linspace(-0.2, 1.2, 15)
y = np.linspace(-0.2, 1.2, 15)
z = np.linspace(-0.2, 1.2, 15)
X, Y, Z = np.meshgrid(x, y, z, indexing='ij')
grid1 = pv.StructuredGrid(X, Y, Z)

# Associate some sinteresting data
P = rng.random(size, dtype=np.float32)
grid1.cell_arrays["P"] = P.flatten()

N = 250
origin = np.array([0.0, 0.5, 0.5])
nvec = np.array([1.0, 0.0, 0.0])
for i in range(N):
    x0 = origin + (i/(N-1))*nvec
    print(f"Generating slice {i}")
    slice_once(grid1, x0, nvec)

```
The above program manage to generate 183 screenshots before it segfaults on my system.

**Screenshots**
No...

-----

**System Information:**
 Please run the following code wherever you are experiencing the bug and paste the output below. This report helps us track down bugs and it is critical to addressing your bug:

```py
# Get system info
import pyvista as pv
print(pv.Report())
```

```txt
--------------------------------------------------------------------------------
  Date: Mon Feb 15 09:30:17 2021 CET

                OS : Linux
            CPU(s) : 16
           Machine : x86_64
      Architecture : 64bit
       Environment : Python
        GPU Vendor : X.Org
      GPU Renderer : Radeon RX 580 Series (POLARIS10, DRM 3.35.0, 5.4.0-65-generic, LLVM 11.0.0)
       GPU Version : 4.6 (Core Profile) Mesa 20.2.6

  Python 3.8.5 (default, Jul 28 2020, 12:59:40)  [GCC 9.3.0]

           pyvista : 0.28.1
               vtk : 9.0.1
             numpy : 1.20.1
           imageio : 2.9.0
           appdirs : 1.4.4
            scooby : 0.5.6
            meshio : 4.3.8
        matplotlib : 3.3.4
--------------------------------------------------------------------------------

```

**Comment by tkoyama010** [comment]
It seems duplicate with #185 . As you're running VTK 9.0 on Linux, this will occur as there's a bug with VTK not closing down the windows/figures even though they're in the background. Fixing vtk version to 8.1.2 may work.
https://github.com/pyvista/pyvista/blob/fbb8570ebd942cdb82a0602daf78a8dff9e2c972/requirements_docs.txt#L18

**Comment by akaszynski** [comment]
@tkoyama010 is exactly right.  This is a known issue of `vtk==9.0.1`.  I'm sure they'll have it fixed in the next release, but that might take a while since VTK wheels tend to take a bit to release.

**Comment by hakostra** [comment]
8.1.2 is so old that there are no wheels for Python 3.8... So that is not an option. Then it's just to be patient. Thanks.

**Comment by hbgtjxzbbx** [comment]
> @tkoyama010 is exactly right. This is a known issue of `vtk==9.0.1`. I'm sure they'll have it fixed in the next release, but that might take a while since VTK wheels tend to take a bit to release.

I met the same error after saving around 100+ plots. I  tried to built the latest vtk from the source but seems the bug hasn't been fixed. Is there any way to workaround?  

Besides, the vtk==8.2.0 doesn't have this issue, but strangely, the plots show and close automatically even with screen_off =True.

**Comment by dgriffiths3** [comment]
@akaszynski you mention this is a known issue from VTK. Do you know of a link to the issue on their GitLab? I can't find anything, but would like to know the status of this. Thanks!

**Comment by akaszynski** [comment]
https://gitlab.kitware.com/vtk/vtk/-/merge_requests/7095

This was merged and dev wheels don't have the same issue.  However, the connection still sticks around and it's easy to go over the default 256 client limit.  We've implemented a workaround for non-interactive plots by using the generic window interactor, but interactive windows will still have this issue.

**Comment by dgriffiths3** [comment]
I'm not doing anything interactive (all remote, `off_screen=True`). I just did a quick update to 0.32.1 and it looks like the problem is solved for me. Thanks a lot!

**Comment by 4ndr3aR** [comment]
I'm at 0.44.2 and it looks like I have this problem.

```
ERROR:root:bad X server connection. DISPLAY=localhost:15.0. Aborting.
Maximum number of clients reached
2025-04-03 19:20:45.369 (  10.296s) [    7F4A873E7080]vtkXOpenGLRenderWindow.:456    ERR| vtkXOpenGLRenderWindow (0xe6e7da0): bad X server connection. DISPLAY=localhost:15.0. Aborting.
```

**Comment by user27182** [comment]
> I'm at 0.44.2 and it looks like I have this problem (running both inside and outside a `screen`, both exporting `DISPLAY=:0` or not).
> 
> ```
> ERROR:root:bad X server connection. DISPLAY=localhost:15.0. Aborting.
> Maximum number of clients reached
> 2025-04-03 19:20:45.369 (  10.296s) [    7F4A873E7080]vtkXOpenGLRenderWindow.:456    ERR| vtkXOpenGLRenderWindow (0xe6e7da0): bad X server connection. DISPLAY=localhost:15.0. Aborting.
> ```

What version of vtk is installed? Can you run
``` python
# Get system info
import pyvista as pv
print(pv.Report())
```

**Comment by 4ndr3aR** [comment]
```
--------------------------------------------------------------------------------
  Date: Thu Apr 03 19:33:39 2025 CEST

                  OS : Linux (Ubuntu 24.04)
              CPU(s) : 48
             Machine : x86_64
        Architecture : 64bit
                 RAM : 125.3 GiB
         Environment : Python
         File system : btrfs
          GPU Vendor : NVIDIA Corporation
        GPU Renderer : NVIDIA RTX A6000/PCIe/SSE2
         GPU Version : 4.5.0 NVIDIA 550.127.05
    MathText Support : True

  Python 3.12.3 (main, Sep 11 2024, 14:17:37) [GCC 13.2.0]

             pyvista : 0.44.2
                 vtk : 9.4.0
               numpy : 2.2.2
          matplotlib : 3.9.2
              scooby : 0.10.0
               pooch : 1.8.2
              pillow : 10.4.0
             imageio : 2.36.0
             IPython : 8.29.0
            colorcet : 3.1.0
             cmocean : 4.0.3
          ipywidgets : 8.1.5
               scipy : 1.14.1
                tqdm : 4.66.6
              meshio : 4.4.6
          jupyterlab : 4.1.8
               trame : 3.8.1
        trame_client : 3.6.1
        trame_server : 3.4.0
           trame_vtk : 2.8.15
       trame_vuetify : 2.9.0
jupyter_server_proxy : 4.4.0
        nest_asyncio : 1.6.0
--------------------------------------------------------------------------------
```

**Comment by 4ndr3aR** [comment]
Well, I'm using GNU parallel to take screenshots of different views of vtk files, so I think I'm passing the limit of 256 connections very quickly.

**Comment by user27182** [comment]
Thank you. Looks like you have vtk 9.4 installed which is not yet supported by PyVista. 

> ```
>              pyvista : 0.44.2
>                  vtk : 9.4.0
> ```

I am not sure if this is the same issue reported above, but may be related to vtk 9.4 specifically. If you upgrade to `vtk==9.4.2`, does the issue persist?

We are having problems trying to make PyVista compatible with VTK 9.4. With `vtk==9.4.2` we are seeing the same warning emitted when building our documentation, see Issue 2 here https://github.com/pyvista/pyvista/pull/6879#issuecomment-2773862865

This _could_ be the root cause of why our documentation build is crashing (which is blocking us from upgrading vtk to 9.4).

Are you able to provide a minimally reproducible example for generating this warning? This would help us a lot with identifying the root cause.

**Comment by 4ndr3aR** [comment]
Sure, here you go.

```
#!/usr/bin/env python3

import pyvista as pv

import subprocess

import argparse

import time

import cv2

from pathlib import Path

def start_plotting(filename, debug_show=False):
	p = pv.Plotter(off_screen=(not debug_show))

	# Read filename
	p.add_mesh(pv.read(filename))

	# Set camera
	p.camera_position = 'xz'

	# Set background
	p.background_color = 'k'

	# Set window size
	p.window_size = (1000, 1000)

	# Set window title
	p.title = f'{filename}'

	# Remove legend
	p.remove_legend()
	# Remove scalar bar
	p.remove_scalar_bar()

	# Show
	if debug_show:
		p.show()

	return p

def rotate(p, elevation, azimuth):
	p.camera.elevation = elevation
	p.camera.azimuth = azimuth

def screenshot(p, out_fn, ext='jpg', quality=90):
	# Take a screenshot
	outf = f'{out_fn}.{ext}'
	img = p.screenshot(filename=None, transparent_background=True)
	if ext == 'jpg':
		img = img[:,:,:3]
		cv2.imwrite(outf, img, [int(cv2.IMWRITE_JPEG_QUALITY), quality])
	else:
		cv2.imwrite(outf, img)

def get_class_and_id_from_filename(filename):
	# 1768-60-6msg_32:TA:r_model1.vtk
	# 2043-81-6qc2_24:X:B8_model1.vtk
	# The first field is the id, the second one is the class
	# Separator is '-'
	fn = str(Path(filename).stem)
	cls = fn.split('-')[1]
	_id = fn.split('-')[0]
	return cls, _id

if __name__ == '__main__':
	parser = argparse.ArgumentParser()

	parser.add_argument(
		'filename', type=str, help='filename to plot',
	)

	# Add debug_show optional and non-positional parameter
	parser.add_argument(
		'--debug_show', action='store_true', help='show plot',
	)

	# Add output directory --out_dir optional and non-positional parameter
	parser.add_argument(
		'--out_dir', type=str, help='output directory', default='/tmp/proteins-screenshots'
	)

	args = parser.parse_args()

	filename = args.filename
	cls, _id = get_class_and_id_from_filename(filename)
	out_dir  = Path(args.out_dir) / cls
	print(f'{_id} - {cls} - Creating directory: {out_dir}')
	Path(out_dir).mkdir(parents=True, exist_ok=True)

	for elevation in range(0, 360, 45):
		for azimuth in range(0, 360, 45):
			p = start_plotting(filename, args.debug_show)
			rotate(p, elevation, azimuth)
			out_fn = Path(out_dir) / f'{Path(filename).stem}-{elevation}-{azimuth}'
			screenshot(p, out_fn, ext='jpg', quality=90)

```

**Comment by user27182** [comment]
Thanks. I am not sure if this is the fix but with headless environments you may need to configure the rendering window with an environment variable:
https://docs.vtk.org/en/latest/advanced/runtime_settings.html#opengl

e.g.
```
export VTK_DEFAULT_OPENGL_WINDOW=vtkOSOpenGLRenderWindow
```

Then check with:

``` python
import vtk
ren_win = vtk.vtkRenderWindow()
print(ren_win.GetClassName())  # Should output vtkOSOpenGLRenderWindow
```

I _think_ your issue is that `vtkXOpenGLRenderWindow` is being used by default (you can check with code above). `vtkXOpenGLRenderWindow` is generating the warning, suggesting that this is indeed the backend being used.

The odd thing is that the source code for this warning suggests that it is only emitted for on-screen plotting. 
https://gitlab.kitware.com/vtk/vtk/-/blob/master/Rendering/OpenGL2/vtkXOpenGLRenderWindow.cxx#L1323-1335 
But I suspect you're getting this in an off-screen context?

**Comment by 4ndr3aR** [comment]
It worked, thanks!

Just added:

```
import os
os.environ['VTK_DEFAULT_OPENGL_WINDOW'] = 'vtkOSOpenGLRenderWindow'
#os.environ['VTK_DEFAULT_OPENGL_WINDOW'] = 'vtkEGLRenderWindow'

import vtk 
ren_win = vtk.vtkRenderWindow()
print(ren_win.GetClassName())  # Should output vtkOSOpenGLRenderWindow
```

At first it wasn't working because I was still at `vtk==9.3.1` , but after upgrading to `9.4.2` it switched renderer to `vtkOSOpenGLRenderWindow`  without problems. It's running with `parallel -j16` and it has already produced more images than a whole night with `vtkXOpenGLRenderWindow` running with  `parallel -j4` (that gave a lot of errors nevertheless).

Thanks again!

**Comment by user27182** [comment]
Glad it worked! Yes this fix is for vtk >= 9.4.
Since this fix requires a non-obvious setting of an environment variable, I'm thinking we should probably emit a warning when using off-screen plotting, and provide a convenience API to easily modify `VTK_DEFAULT_OPENGL_WINDOW`.
